core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+        stagesByFragmentId = stages.stream()
+                .collect(toImmutableMap(stage -> stage.getFragment().getId(), identity()));
+        extractDependencies(
+                stages.stream()
+                        .map(StageExecution::getFragment)
+                        .collect(toImmutableList())).stream()
+                // start non-lazy stages
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+        // start stages without any dependencies
+        fragmentDependency.vertexSet().stream()
+                .filter(fragmentId -> fragmentDependency.inDegreeOf(fragmentId) == 0)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+    }
+
+    @Override
+    public Set<PipelinedStageExecution> getStagesToSchedule()
+    {
+        schedule();
+        return activeSources.stream()
+                .map(StageExecution::getPipelinedStageExecution)
+                .collect(toImmutableSet());
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        // dependency graph contains both running and not started fragments
+        return fragmentDependency.vertexSet().isEmpty();
+    }
+
+    @Override
+    public synchronized ListenableFuture<Void> getRescheduleFuture()
+    {
+        return rescheduleFuture;
+    }
+
+    @VisibleForTesting
+    void schedule()
+    {
+        removeCompletedStages();
+        unblockStages();
+    }
+
+    @VisibleForTesting
+    DirectedGraph<PlanFragmentId, FragmentsEdge> getFragmentDependency()
+    {
+        return fragmentDependency;
+    }
+
+    @VisibleForTesting
+    Set<StageExecution> getActiveSources()
+    {
+        return activeSources;
+    }
+
+    private void removeCompletedStages()
+    {
+        Set<StageExecution> completedStages = activeSources.stream()
+                .filter(this::isStageCompleted)
+                .collect(toImmutableSet());
+        // remove completed stages outside of Java stream to prevent concurrent modification
+        completedStages.forEach(this::removeCompletedStage);
+    }
+
+    private void removeCompletedStage(StageExecution stage)
+    {
+        // start all stages that depend on completed stage
+        PlanFragmentId fragmentId = stage.getFragment().getId();
+        fragmentDependency.outgoingEdgesOf(fragmentId).stream()
+                .map(FragmentsEdge::getTarget)
+                // filter stages that depend on completed stage only
+                .filter(dependentFragmentId -> fragmentDependency.inDegreeOf(dependentFragmentId) == 1)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+        fragmentDependency.removeVertex(fragmentId);
+        fragmentTopology.removeVertex(fragmentId);
+        activeSources.remove(stage);
+    }
+
+    private void unblockStages()
+    {
+        // find stages that are blocked on full task output buffer
```

Comment:
does `isAnyTaskBlocked` imply that reason for task being blocked is that output buffer is full?

Reply:
The name doesn't. It feels like we should rename it.